### PR TITLE
Add requestToken endpoints

### PR DESCRIPTION
**Pull request by dbkr**
Implement https://github.com/matrix-org/matrix-doc/pull/346/files

### CODE REVIEW

File: synapse/rest/client/v2_alpha/account.py [hunk 1]
```diff
@@ -36,11 +36,16 @@ def __init__(self, hs):
         self.hs = hs
         self.auth = hs.get_auth()
         self.auth_handler = hs.get_auth_handler()
+        self.identity_handler = hs.get_handlers().identity_handler
 
     @defer.inlineCallbacks
     def on_POST(self, request):
         yield run_on_reactor()
 
+        if '/account/password/email/requestToken' in request.path:
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: Why aren't these just separate handlers?

File: synapse/rest/client/v2_alpha/account.py [hunk 2]
```diff
@@ -155,6 +187,33 @@ def on_POST(self, request):
...
+                absent.append(k)
+
+        if len(absent) > 0:
+            raise SynapseError(400, "Missing params: %r" % absent, Codes.MISSING_PARAM)
+
+        existingUid = yield self.hs.get_datastore().get_user_id_by_threepid(
+            'email', body['email']
+        )
+
+        logger.error("existing %r", existingUid)
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: Left over debug logs?

**Comment by dbkr** [comment]
ptal

File: synapse/rest/client/v2_alpha/account.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: Is this a thing?

File: synapse/rest/client/v2_alpha/account.py [hunk 3]
```diff
@@ -28,8 +28,43 @@
...
+    def on_POST(self, request):
+        body = parse_json_object_from_request(request)
+
+        required = ['id_server', 'client_secret', 'email', 'send_attempt']
+        absent = []
+        for k in required:
+            if k not in body:
+                absent.append(k)
+
+        if len(absent) > 0:
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: Nit: Its more pythonic to just do `if absent:`

File: synapse/rest/client/v2_alpha/account.py [hunk 4]
```diff
@@ -28,8 +28,43 @@
...
+        )
+
+        if existingUid is None:
+            raise SynapseError(400, "Email not found", Codes.THREEPID_NOT_FOUND)
+
+        ret = yield self.identity_handler.requestEmailToken(**body)
+        defer.returnValue((200, ret))
+
+    def on_OPTIONS(self, _):
+        return 200, {}
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: `on_OPTIONS` should be handled automatically?

File: synapse/rest/client/v2_alpha/register.py [hunk 5]
```diff
@@ -41,8 +41,43 @@ def compare_digest(a, b):
...
+            raise SynapseError(400, "Missing params: %r" % absent, Codes.MISSING_PARAM)
+
+        existingUid = yield self.hs.get_datastore().get_user_id_by_threepid(
+            'email', body['email']
+        )
+
+        if existingUid is not None:
+            raise SynapseError(400, "Email is already in use", Codes.THREEPID_IN_USE)
+
+        ret = yield self.identity_handler.requestEmailToken(**body)
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/register.py]: Does this not explode if there are extra keys in the body?

File: synapse/rest/client/v2_alpha/account.py (hunk 4 above)
Comment (dbkr) [inline synapse/rest/client/v2_alpha/account.py]: Ah, good to know, just copied it from the other handlers.

File: synapse/rest/client/v2_alpha/register.py (hunk 5 above)
Comment (dbkr) [inline synapse/rest/client/v2_alpha/register.py]: requestEmailToken takes **kwargs so it shouldn't do (next_link is passed through this way)

**Comment by erikjohnston** [comment]
LGTM

**Comment by erikjohnston** [comment]
(Is this sytest'ed or even sytest'able?)

**Comment by dbkr** [comment]
It would be tricky to sytest: currently we could test if it rejected appropriately, but since we allow the HS to HTTP 200 the request and tell the user by email that the email is already taken / not found, the tests would then fail if an HS did that.

**Pull request merged**